lib/pengine/utils.c

```diff
@@ -2080,57 +2080,129 @@ rsc_action_digest_cmp(resource_t * rsc, xmlNode * xml_op, node_t * node,
     return data;
 }
 
+/*!
+ * \internal
+ * \brief Create an unfencing summary for use in special node attribute
+ *
+ * Create a string combining a fence device's resource ID, agent type, and
+ * parameter digest (whether for all parameters or just non-private parameters).
+ * This can be stored in a special node attribute, allowing us to detect changes
+ * in either the agent type or parameters, to know whether unfencing must be
+ * redone or can be safely skipped when the device's history is cleaned.
+ *
+ * \param[in] rsc_id        Fence device resource ID
+ * \param[in] agent_type    Fence device agent
+ * \param[in] param_digest  Fence device parameter digest
+ *
+ * \return Newly allocated string with unfencing digest
+ * \note The caller is responsible for freeing the result.
+ */
+static inline char *
+create_unfencing_summary(const char *rsc_id, const char *agent_type,
+                         const char *param_digest)
+{
+    return crm_strdup_printf("%s:%s:%s", rsc_id, agent_type, param_digest);
+}
+
+/*!
+ * \internal
+ * \brief Check whether a node can skip unfencing
+ *
+ * Check whether a fence device's current definition matches a node's
+ * stored summary of when it was last unfenced by the device.
+ *
+ * \param[in] rsc_id        Fence device's resource ID
+ * \param[in] agent         Fence device's agent type
+ * \param[in] digest_calc   Fence device's current parameter digest
+ * \param[in] node_summary  Value of node's special unfencing node attribute
+ *
+ * \return TRUE if digest matches, FALSE otherwise
+ */
+static bool
+unfencing_digest_matches(const char *rsc_id, const char *agent,
+                         const char *digest_calc, const char *node_summary)
+{
+    bool matches = FALSE;
+
+    if (rsc_id && agent && digest_calc && node_summary) {
+        char *search_secure = create_unfencing_summary(rsc_id, agent,
+                                                       digest_calc);
+
+        matches = (strstr(node_summary, search_secure) != NULL);
```

Comment:
I see `strstr` approach is pre-existing, but I am lost on purpose of that.
Some kind of extensibility going forward?

Reply:
The node summary is a comma-separated list of unfencing summaries for all capable fence devices, so strstr() checks whether it's in there somewhere.

Reply:
Ah, I was missing the very non-apparent link/data-flow/round-trip
over attrd, if I grok that right.

Anyway, what I saw as problematic is the fact that this method can fail
easily with user-provided `rsc->id` (they can use the fixed sufix and
gradually extended prefixes), i.e., it's too naive.

Solutions are plentiful, e.g,:

```
if ((ptr = strstr(haystack, needle)) != NULL
		&& (ptr == haystack || haystack[ptr - haystack - 1] == ',')) {
	/* success */
}
```

Reply:
Nope, unfencing summaries are <resource-id>:<agent-name>:<hash>. Colons are not allowed in resource IDs, so the only possible collision would be a different agent name that exactly matches that entire string -- not a real risk, and it would break the parsing already anyway.

Reply:
Also, information that there's a list of of unfencing summaries
could possibly be noted in `node_summary` parameter's doc.
`node's special unfencing node attribute` keeps readers at guess
while it's a crucial implementation detail here.

Reply:
I don't understand what you mean with "entire string".

Reply:
I think that under some circumstances, you can have /agent-name/:/hash/ part
the same amongst two instances, and /resource-id/ of one be the suffix of the other.
In a wrong combination, it could cause troubles, couldn't it?

Reply:
I see what you're saying. That could only happen if the identical agent and configuration parameters were used with multiple device names. That's unlikely but possible, so probably worth taking care of.

Reply:
I see it can be considered (in a sense) unfair approach to reverse
the flow to make up for (seemingly artificial) problems intentionally, 
but that's exactly how odds/entropy are playing against us when there's
a slight chance.  Thanks!

Reply:
Upon further investigation, the calculated digest includes the device ID and agent, so there is no chance of collision.

Reply:
I've tried to track that fact down, but I failed, the data flow is
extremely difficult to follow.  I guess it stems from the fact that
whole chunk of XML gets, at an occasion, serialized (so presumably
with ID included), and MD5 checksum is calculated for that.

Reply:
"Extremely difficult to follow" is an understatement :)

The hash isn't of the resource XML directly, but of a stringified combination of the resource's instance attributes hash table and meta-attributes hash-table (and potentially other stuff). The resource element's XML attributes (id and agent info) are added to the resource's meta-attributes hash table when unpacking.